## Programa Ingenias+ Data Science

## Trabajando con Pandas

1) Importa las librerias requeridas: `Pandas` y `NumPy`.  

Una buena practica a la hora de escribir codigo es de comentar siempre que estoy haciendo (usando `#`) en la linea anterior. Por mas obvio que parezca lo que voy a hacer, puede ser que tenga que compartir el documento con otras personas y no sepan que quise hacer en cada linea. O que yo mismo lea el codigo tiempo despues y no recuerde porque lo hice.

In [2]:
import pandas as pd
import numpy as np


2) Lee el archivo csv `life_expectancy_data` usando `pandas`. Guardalo en una variable llamada `expvida`.

In [6]:
# Leo los datos en un DataFrame
path= '/content/life_expectancy_data.csv'
df= pd.read_csv(path)

3) ¿Que tipo de estructura de datos contiene la variable `expvida`?

In [7]:
# Obtengo el tipo de estructura de datos
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

4) ¿Cuantas filas y columnas tiene `expvida`?

In [9]:
# Uso .shape para obtener cantidad de filas y columnas
df.shape


(2938, 22)

**Numero de filas**: 2938

**Numero de columnas**:22

5) ¿Cual es el nombre de las columnas contenidas en `expvida`?

In [10]:
# Obtengo el nombre de las columnas
df.columns


Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

6) Inspecciona las primeras 10 filas de `expvida`

In [11]:
# Inspecciono las primeras filas
df.head(10)


Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   
5  Afghanistan  2010  Developing              58.8            279.0   
6  Afghanistan  2009  Developing              58.6            281.0   
7  Afghanistan  2008  Developing              58.1            287.0   
8  Afghanistan  2007  Developing              57.5            295.0   
9  Afghanistan  2006  Developing              57.3            295.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582         62.0       492  ...   
2             66     0.01               73.219243         64.0       430  ...   
3             69     0.01               78.184215         67.0      2787  ...   
4             71     0.01                7.097109         68.0      3013  ...   
5             74     0.01               79.679367         66.0      1989  ...   
6             77     0.01               56.762217         63.0      2861  ...   
7             80     0.03               25.873925         64.0      1599  ...   
8             82     0.02               10.910156         63.0      1141  ...   
9             84     0.03               17.171518         64.0      1990  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0               8.16         65.0        0.1  584.259210  33736494.0   
1   58.0               8.18         62.0        0.1  612.696514    327582.0   
2   62.0               8.13         64.0        0.1  631.744976  31731688.0   
3   67.0               8.52         67.0        0.1  669.959000   3696958.0   
4   68.0               7.87         68.0        0.1   63.537231   2978599.0   
5   66.0               9.20         66.0        0.1  553.328940   2883167.0   
6   63.0               9.42         63.0        0.1  445.893298    284331.0   
7   64.0               8.33         64.0        0.1  373.361116   2729431.0   
8   63.0               6.73         63.0        0.1  369.835796  26616792.0   
9   58.0               7.43         58.0        0.1  272.563770   2589345.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   
5                   18.4                 18.4   
6                   18.6                 18.7   
7                   18.8                 18.9   
8                   19.0                 19.1   
9                   19.2                 19.3   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  
5                            0.448        9.2  
6                            0.434        8.9  
7                            0.433        8.7  
8                            0.415        8.4  
9                            0.405        8.1  

[10 rows x 22 columns]

7) Y ahora las 10 ultimas.

In [12]:
# Inspecciono las ultimas filas
df.tail(10)


Country  Year      Status  Life expectancy   Adult Mortality  \
2928  Zimbabwe  2009  Developing              50.0            587.0   
2929  Zimbabwe  2008  Developing              48.2            632.0   
2930  Zimbabwe  2007  Developing              46.6             67.0   
2931  Zimbabwe  2006  Developing              45.4              7.0   
2932  Zimbabwe  2005  Developing              44.6            717.0   
2933  Zimbabwe  2004  Developing              44.3            723.0   
2934  Zimbabwe  2003  Developing              44.5            715.0   
2935  Zimbabwe  2002  Developing              44.8             73.0   
2936  Zimbabwe  2001  Developing              45.3            686.0   
2937  Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
2928             30     4.64                1.040021         73.0       853   
2929             30     3.56               20.843429         75.0         0   
2930             29     3.88               29.814566         72.0       242   
2931             28     4.57               34.262169         68.0       212   
2932             28     4.14                8.717409         65.0       420   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
2928  ...   69.0               6.26         73.0       18.1   65.824121   
2929  ...   75.0               4.96         75.0       20.5  325.678573   
2930  ...   73.0               4.47         73.0       23.7  396.998217   
2931  ...   71.0               5.12          7.0       26.8  414.796232   
2932  ...   69.0               6.44         68.0       30.3  444.765750   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
2928   1381599.0                    7.5                  7.4   
2929  13558469.0                    7.8                  7.8   
2930   1332999.0                    8.2                  8.2   
2931  13124267.0                    8.6                  8.6   
2932    129432.0                    9.0                  9.0   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
2928                            0.419        9.9  
2929                            0.421        9.7  
2930                            0.414        9.6  
2931                            0.408        9.5  
2932                            0.406        9.3  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427       10.0  
2936                            0.427        9.8  
2937                            0.434        9.8  

[10 rows x 22 columns]

8) ¿Que tipos de datos contiene cada una de las columnas de `expvida`?

In [13]:
# Tipo de datos en cada columna
df.dtypes


Country                             object
Year                                 int64
Status                              object
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
 BMI                               float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: object

9) ¿Hay valores faltantes en alguna de las columnas de `expvida`?

In [16]:
# Obtengo la cantidad de datos faltantes en cada columna
df.isnull().sum()

Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64

10) Selecciona el maximo de datos faltantes. Calcula que porcentaje de los datos representan los datos faltantes.

In [17]:
# Obtengo el maximo valor de datos faltantes y lo guardo en la variable masnan
masnan= max(df.isnull().sum())
masnan


652

In [23]:
# Obtengo la cantidad de filas total y divido el maximo valor de datos faltantes por este numero
# y lo multiplico por 100
filas= df.shape[0]
faltantes= max(df.isnull().sum())
total= (faltantes/filas) * 100
total


22.19196732471069

In [24]:
# Imprimo el maximo porcentaje de datos faltantes. Uso f-strings.
print(f'Porcentaje máximo de datos faltantes: {total}')


Porcentaje máximo de datos faltantes: 22.19196732471069


- str.format()
- Template Class
- f-strings

#### Ahora vamos a realizar algunos calculos y manipulaciones sobre las columnas.

Una función útil para este fin es: `.rename()`. Y podemos utilizarla de la siguiente manera: Si `df` es mi DataFrame y quiero renombrar la columna `columna` como `columna1`, y la columna `columna2` como `columna3`, utilizo la siguiente sintaxis:
  
`df.rename(columns={"columna": "columna1", "columna2": "columna3"}, inplace=True)`

11) Renombra la columna 'Life expectancy ' como 'life_expectancy', ' BMI ' como 'bmi' y ''Measles ' como 'measles'. No olvides de usar `inplace` para modificar el DataFrame original. Luego chequea que el cambio se haya hecho correctamente.

In [29]:
# Renombro columnas del DataFrame
df= df.rename(columns= {'Life expectancy ':'life_expectancy', ' BMI ': 'bmi', 'Measles ': 'measles'})


In [30]:
# Chequeo que se hayan reemplazado correctamente
df.columns

Index(['Country', 'Year', 'Status', 'life_expectancy', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'measles', 'bmi', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

12) ¿Cual es el promedio de expectativa de vida (Life expectancy) en general?

In [31]:
# Obtengo el promedio de la columna
df['life_expectancy']. mean()


69.22493169398908

13) ¿Cual es el promedio de expectativa de vida (Life expectancy) por país?

In [34]:
# Obtengo el promedio por pais
df.groupby('Country')['life_expectancy'].mean()


Country
Afghanistan                           58.19375
Albania                               75.15625
Algeria                               73.61875
Angola                                49.01875
Antigua and Barbuda                   75.05625
                                        ...   
Venezuela (Bolivarian Republic of)    73.38750
Viet Nam                              74.77500
Yemen                                 63.86250
Zambia                                53.90625
Zimbabwe                              50.48750
Name: life_expectancy, Length: 193, dtype: float64

14) ¿Cual es la expectativa de vida mas alta y mas baja? Usa la funcion de pandas y la funcion de NumPy y compara el resultado. ¿Que paises tienen esas expectativas de vida?

In [35]:
# Maxima expectativa de vida usando numpy
np.max(df['life_expectancy'])


89.0

In [38]:
# Maxima expectativa de vida usando pandas
df['life_expectancy'].max()


89.0

In [39]:
# Minima expectativa de vida usando numpy
np.min(df['life_expectancy'])


36.3

In [40]:
# Minima expectativa de vida usando pandas
df['life_expectancy'].min()

36.3

In [43]:
# Muestro los resultados maximos
numpy_max= np.max(df['life_expectancy'])
pandas_max=df['life_expectancy'].max()
print(f'Máximo Numpy:{numpy_max}')
print(f'Máximo Pandas:{pandas_max}')


Máximo Numpy:89.0
Máximo Pandas:89.0


In [44]:
# Muestro los resultados minimos
numpy_min= np.min(df['life_expectancy'])
pandas_min=df['life_expectancy'].min()
print(f'Mínimo Numpy:{numpy_min}')
print(f'Mínimo Pandas:{pandas_min}')

Mínimo Numpy:36.3
Mínimo Pandas:36.3


In [45]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al maximo
df[df['life_expectancy'] == pandas_max]


Country  Year      Status  life_expectancy  Adult Mortality  \
241       Belgium  2014   Developed             89.0             76.0   
915       Finland  2014  Developing             89.0             78.0   
937        France  2008  Developing             89.0             88.0   
938        France  2007  Developing             89.0             89.0   
995       Germany  2014   Developed             89.0             69.0   
1293        Italy  2004   Developed             89.0             66.0   
1850  New Zealand  2010   Developed             89.0             71.0   
1916       Norway  2009   Developed             89.0             67.0   
2056     Portugal  2014   Developed             89.0             78.0   
2433        Spain  2007   Developed             89.0             72.0   
2513       Sweden  2007   Developed             89.0             63.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  measles  \
241               0    12.60             7163.348923         98.0       70   
915               0     8.80             6164.455402          NaN        0   
937               3    11.90             7002.785925         47.0      604   
938               3    12.20               64.737149         42.0       39   
995               2    11.03              941.756291         88.0      443   
1293              2     8.98             4270.914882         96.0      599   
1850              0     9.62             6664.148708          9.0       43   
1916              0     6.68              142.374943          NaN        2   
2056              0     9.88              271.254553         98.0        0   
2433              2    11.05              510.932701         96.0      267   
2513              0     6.90             7593.391622          NaN        1   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS          GDP  \
241   ...   99.0               1.59         99.0        0.1  47439.39684   
915   ...   98.0               9.68         98.0        0.1  49914.61864   
937   ...   98.0               1.57         98.0        0.1  45413.65710   
938   ...   99.0               1.45         98.0        0.1    416.58397   
995   ...   94.0              11.30         95.0        0.1   4792.65288   
1293  ...   97.0               8.49         94.0        0.1  31174.56118   
1850  ...   93.0              11.20         93.0        0.1  33691.34837   
1916  ...   94.0               9.49         94.0        0.1    817.77681   
2056  ...   98.0               9.50         98.0        0.1   2277.53613   
2433  ...   96.0               8.36         96.0        0.1   3279.41400   
2513  ...   98.0               8.92         98.0        0.1  53324.37937   

      Population   thinness  1-19 years   thinness 5-9 years  \
241     112957.0                    1.0                  1.0   
915    5461512.0                    0.9                  0.8   
937    6437499.0                    0.6                  0.6   
938    6416229.0                    0.6                  0.6   
995      89825.0                    1.1                  1.1   
1293  57685327.0                    0.5                  0.5   
1850         NaN                    0.3                  0.3   
1916   4828726.0                    0.7                  0.7   
2056     14162.0                    0.7                  0.5   
2433   4522683.0                    0.6                  0.5   
2513    914892.0                    1.3                  1.3   

      Income composition of resources  Schooling  
241                             0.890       16.3  
915                             0.890       17.0  
937                             0.877       16.1  
938                             0.873       16.1  
995                             0.920       17.0  
1293                            0.846       15.6  
1850                            0.899       20.3  
1916                            0.936       17.4  
2056                            0.837       16.8  
2433

In [46]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al minimo
df[df['life_expectancy'] == pandas_min]

Country  Year      Status  life_expectancy  Adult Mortality  \
1127   Haiti  2010  Developing             36.3            682.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  measles  \
1127             23     5.76               36.292918          NaN        0   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
1127  ...   66.0                8.9         66.0        1.9  662.279518   

      Population   thinness  1-19 years   thinness 5-9 years  \
1127   9999617.0                    4.0                  4.0   

      Income composition of resources  Schooling  
1127                             0.47        8.6  

[1 rows x 22 columns]

15) ¿Cuantas categorias encontramos en la columna "Status"? _**Pista**_: Lee acerca de la función [unique()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html?highlight=unique#pandas.Series.unique)

In [47]:
# Obtengo los valores unicos de la columna status
df['Status'].unique()

array(['Developing', 'Developed'], dtype=object)

16) ¿Cuantos paises se ubican en cada una de esas categorias de "Status"? _**Pista**_: Lee acerca de la función [value_counts()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html)

In [51]:
# Cuento cuantos valores hay de cada categoria
df.groupby('Country')['Status'].value_counts()

Country                             Status    
Afghanistan                         Developing    16
Albania                             Developing    16
Algeria                             Developing    16
Angola                              Developing    16
Antigua and Barbuda                 Developing    16
                                                  ..
Venezuela (Bolivarian Republic of)  Developing    16
Viet Nam                            Developing    16
Yemen                               Developing    16
Zambia                              Developing    16
Zimbabwe                            Developing    16
Name: count, Length: 193, dtype: int64

18) ¿Cuantos paises presentaron algun caso de Sarampion (**Measles**) en el 2015? **_Pista_**: Lee acerca de la función [.count()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.count.html?highlight=count#pandas.DataFrame.count)

In [61]:
# Cuento cuantas filas tienen valores distintos de cero
filter= df[df['Year']== 2015]
filter_count= filter[filter['measles'] != 0]
filter_count['Country'].count()


121